pyccel/codegen/printing/cucode.py

```diff
@@ -0,0 +1,86 @@
+# coding: utf-8
+#------------------------------------------------------------------------------------------#
+# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
+# go to https://github.com/pyccel/pyccel/blob/master/LICENSE for full license details.     #
+#------------------------------------------------------------------------------------------#
+"""
+Provide tools for generating and handling CUDA code.
+This module is designed to interface Pyccel's Abstract Syntax Tree (AST) with CUDA,
+enabling the direct translation of high-level Pyccel expressions into CUDA code.
+"""
+
+from pyccel.codegen.printing.ccode import CCodePrinter, c_library_headers
+
+from pyccel.ast.core        import Import, Module
+
+from pyccel.errors.errors   import Errors
+
+
+errors = Errors()
+
+# TODO: add examples
+
+__all__ = ["CudaCodePrinter"]
+
+class CudaCodePrinter(CCodePrinter):
+    """
+    Print code in CUDA format.
+
+    This printer converts Pyccel's Abstract Syntax Tree (AST) into strings of CUDA code.
+    Navigation through this file utilizes _print_X functions,
+    as is common with all printers.
+
+    Parameters
+    ----------
+    filename : str
+            The name of the file being pyccelised.
+    prefix_module : str
+            A prefix to be added to the name of the module.
+    """
+    printmethod = "_cucode"
+    language = "cuda"
+
+    def __init__(self, filename, prefix_module = None):
+
+        errors.set_target(filename, 'file')
+
+        super().__init__(filename)
+
+    def _print_Module(self, expr):
+        self.set_scope(expr.scope)
+        self._current_module = expr.name
+        body    = ''.join(self._print(i) for i in expr.body)
+
+        global_variables = ''.join(self._print(d) for d in expr.declarations)
+
+        # Print imports last to be sure that all additional_imports have been collected
+        imports = [Import(expr.name, Module(expr.name,(),())), *self._additional_imports.values()]
+        c_headers_imports = ''
+        local_imports = ''
+
+        for imp in imports:
+            if imp.source in c_library_headers:
+                c_headers_imports += self._print(imp)
+            else:
+                local_imports += self._print(imp)
```

Comment:
Generator expressions are usually much faster than loops
```suggestion
        c_headers_imports = ''.join(self._print(imp) for imp in imports if imp.source in c_library_headers)
        local_imports = ''.join(self._print(imp) for imp in imports if imp.source not in c_library_headers)
```

Reply:
is this considered better even when the suggestion have 2 `for` loops?

Reply:
I thought that the difference was quite large (e.g. factor 10) but it seems that this is not the case. I have made a small test:
```sh
SETUP1="
def func1():
    a = ''.join(str(i) for i in range(1000) if i%2 == 0)
    b = ''.join(str(i) for i in range(1000) if i%2 == 1)
"

SETUP2="def func2():
    a = ''
    b = ''
    for i in range(1000):
        if i%2 == 0:
            a += str(i)
        else:
            b += str(i)
"

python3 -m timeit -s "$SETUP1" "func1()"
python3 -m timeit -s "$SETUP2" "func2()"
```
My results are:
**Python 3.8**
2000 loops, best of 5: 107 usec per loop
2000 loops, best of 5: 101 usec per loop

**Python 3.10**
5000 loops, best of 5: 98.1 usec per loop
5000 loops, best of 5: 85.5 usec per loop

**Python 3.11**
5000 loops, best of 5: 72.1 usec per loop
5000 loops, best of 5: 61.3 usec per loop

**Python 3.12**
5000 loops, best of 5: 84.3 usec per loop
5000 loops, best of 5: 71.5 usec per loop

So with 2 loops and an if block it seems that an explicit loop is slightly faster.

This result does seem to be specific to the if/else block though. In the second case the if only needs checking once whereas with the loop comprehension it is checked twice. If I remove the if condition then the 2 list comprehensions are faster than a loop.

Reply:
oh this is interesting :thinking: good to know.